File: client/src/util/posOrderRequest.js
// Body of POST /pos/orders. Pure function, so the exact request contract can be unit-tested.
//
// The POS has exactly two customer modes and the contract distinguishes them by PRESENCE alone:
//   customerUserId present -> the registered customer the cashier explicitly selected
//   customerUserId absent  -> a genuine walk-in sale (order.user stays NULL)
// Nothing else identifies the customer: customerName/phoneNumber are billing details for the
// receipt, and the backend never uses them to find or associate an account.
//
// Because "absent" is what makes a sale a walk-in, a selected customer carrying no stable userId
// must never be allowed to fall through as absent - that would silently record a registered
// customer's purchase against no account at all. It throws instead, and the caller reports the
// failure rather than billing the sale as a walk-in.
export const buildPosOrderRequest = ({customer = null, customerName = "", phoneNumber = "",
                                         paymentMethod, cartItems}) => {
    if (customer && !customer.userId) {
        throw new Error(
            "This sale is for a registered customer, but that customer's id is missing. "
            + "Select the customer again before taking payment.");
    }
    const name = (customerName || "").trim();
    return {
        ...(customer ? {customerUserId: customer.userId} : {}),
        ...(name ? {customerName: name} : {}),
        ...(phoneNumber ? {phoneNumber} : {}),
        // Only itemId + quantity: name, price, subtotal, tax and grand total are resolved and
        // computed by the backend from its own catalog.
        cartItems: (cartItems || []).map(({itemId, quantity}) => ({itemId, quantity})),
        paymentMethod,
    };
};

// Body of POST /orders. The online customer is the authenticated account, so the request carries
// no identity at all - not even a name or phone number.
export const buildOnlineOrderRequest = ({paymentMethod, cartItems}) => ({
    cartItems: (cartItems || []).map(({itemId, quantity}) => ({itemId, quantity})),
    paymentMethod,
});

File: billingsoftware/src/main/java/in/vedchangani/billingsoftware/io/PosOrderRequest.java
package in.vedchangani.billingsoftware.io;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.Pattern;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.List;

// Request body for POST /pos/orders. There is deliberately no salesChannel, createdBy or
// user-object field: the channel comes from the endpoint and the creator from the authenticated
// principal. customerUserId only *selects* an existing registered customer; the backend resolves
// the actual UserEntity itself. customerName/phoneNumber are billing details for the receipt and
// are never used to find or associate an account.
@Data
@AllArgsConstructor
@NoArgsConstructor
@Builder
public class PosOrderRequest {

    // The stable userId returned by GET /pos/customers. Omit (null) for a walk-in sale.
    private String customerUserId;

    private String customerName;

    @Pattern(regexp = "^[0-9]{10}$", message = "Phone number must be exactly 10 digits")
    private String phoneNumber;

    @NotBlank(message = "Payment method is required")
    private String paymentMethod;

    @NotEmpty(message = "Cart must not be empty")
    @Valid
    private List<OrderRequest.OrderItemRequest> cartItems;
}
